File: app/src/main/java/com/fitc/wifihotspot/TetherManager.java
package com.fitc.wifihotspot;

import android.annotation.SuppressLint;
import android.content.Context;
import android.net.ConnectivityManager;
import android.os.Handler;
import android.util.Log;

import com.android.dx.stock.ProxyBuilder;

import java.io.File;
import java.lang.reflect.Method;
import java.util.Arrays;

/**
 * Created by jonro on 19/03/2018.
 */

public class TetherManager {
    private static final String TAG = TetherManager.class.getSimpleName();

    private final Context mContext;
    private final ConnectivityManager mConnectivityManager;

    public TetherManager(Context context) {
        mContext = context;
        mConnectivityManager = mContext.getSystemService(ConnectivityManager.class);
    }
    /**
     * Checks where tethering is on.
     * This is determined by the getTetheredIfaces() method,
     * that will return an empty array if not devices are tethered
     *
     * @return true if a tethered device is found, false if not found
     */
    public boolean isTetherActive() {
        try {
            @SuppressLint("DiscouragedPrivateApi") Method method = mConnectivityManager.getClass().getDeclaredMethod("getTetheredIfaces");
            String[] res = (String[]) method.invoke(mConnectivityManager);
            Log.d(TAG, "getTetheredIfaces invoked");
            Log.d(TAG, Arrays.toString(res));
            assert res != null;
            if (res.length > 0) {
                return true;
            }
        } catch (Exception e) {
            Log.e(TAG, "Error in getTetheredIfaces");
            e.printStackTrace();
        }
        return false;
    }

    public static class TetherMode {
        public static final int TETHERING_WIFI = 0;
        public static final int TETHERING_USB = 1;
        public static final int TETHERING_BLUETOOTH = 2;
        public static final int TETHERING_NCM = 4;
        public static final int TETHERING_ETHERNET = 5;
        public static final int TETHERING_WIGIG = 6;
    }

    /**
     * This enables tethering using the ssid/password defined in Settings App>Hotspot & tethering
     * Does not require app to have system/privileged access
     * Credit: Vishal Sharma - https://stackoverflow.com/a/52219887
     */
    @SuppressWarnings("UnusedReturnValue")
    public boolean startTethering(int mode, final OnStartTetheringCallback callback) {

        // On Pie if we try to start tethering while it is already on, it will
        // be disabled. This is needed when startTethering() is called programmatically.
        if (isTetherActive()) {
            Log.d(TAG, "Tether already active, returning");
            return false;
        }

        File outputDir = mContext.getCodeCacheDir();
        Object proxy;
        try {
            //noinspection unchecked
            proxy = ProxyBuilder.forClass(OnStartTetheringCallbackClass())
                    .dexCache(outputDir).handler((proxy1, method, args) -> {
                        switch (method.getName()) {
                            case "onTetheringStarted":
                                if (callback != null) {
                                    callback.onTetheringStarted();
                                }
                                break;
                            case "onTetheringFailed":
                                if (callback != null) {
                                    callback.onTetheringFailed();
                                }
                                break;
                            default:
                                ProxyBuilder.callSuper(proxy1, method, args);
                        }
                        //noinspection SuspiciousInvocationHandlerImplementation
                        return null;
                    }).build();
        } catch (Exception e) {
            Log.e(TAG, "Error in enableTethering ProxyBuilder");
            e.printStackTrace();
            return false;
        }

        Method method;
        try {
            method = mConnectivityManager.getClass().getDeclaredMethod("startTethering", int.class, boolean.class, OnStartTetheringCallbackClass(), Handler.class);
            method.invoke(mConnectivityManager, mode, false, proxy, null);
            Log.d(TAG, "startTethering invoked");
            return true;
        } catch (Exception e) {
            Log.e(TAG, "Error in enableTethering");
            e.printStackTrace();
        }
        return false;
    }

    public void stopTethering(int mode) {
        try {
            Method method = mConnectivityManager.getClass().getDeclaredMethod("stopTethering", int.class);
            method.invoke(mConnectivityManager, mode);
            Log.d(TAG, "stopTethering invoked");
        } catch (Exception e) {
            Log.e(TAG, "stopTethering error: " + e);
            e.printStackTrace();
        }
    }

    @SuppressLint("PrivateApi")
    @SuppressWarnings("rawtypes")
    private Class OnStartTetheringCallbackClass() {
        try {
            return Class.forName("android.net.ConnectivityManager$OnStartTetheringCallback");
        } catch (ClassNotFoundException e) {
            Log.e(TAG, "OnStartTetheringCallbackClass error: " + e);
            e.printStackTrace();
        }
        return null;
    }
}

File: app/src/main/java/com/fitc/wifihotspot/TetherManager.kt
@file:Suppress("unused")

package com.fitc.wifihotspot

import android.annotation.SuppressLint
import android.content.Context
import android.net.ConnectivityManager
import android.os.Handler
import android.util.Log
import com.android.dx.stock.ProxyBuilder
import java.io.File

/**
 * Created by jonro on 19/03/2018.
 **/
class TetherManager(private val context: Context) {
    private val logTag = "Telegram-RC.TetherManager"
    private val connectivityManager: ConnectivityManager =
        context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager

    /**
     * Checks where tethering is on.
     * This is determined by the getTetheredIfaces() method,
     * that will return an empty array if not devices are tethered
     *
     * @return true if a tethered device is found, false if not found
     */
    fun isTetherActive(): Boolean {
        return try {
            @SuppressLint("DiscouragedPrivateApi") val method =
                connectivityManager.javaClass.getDeclaredMethod("getTetheredIfaces")
            val res = method.invoke(connectivityManager) as Array<*>
            res.isNotEmpty()
        } catch (e: NoSuchMethodException) {
            Log.e(logTag, "getTetheredIfaces method not found: $e")
            false
        } catch (e: IllegalAccessException) {
            Log.e(logTag, "getTetheredIfaces access denied: $e")
            false
        } catch (e: java.lang.reflect.InvocationTargetException) {
            Log.e(logTag, "getTetheredIfaces invocation failed: $e")
            false
        } catch (e: ClassCastException) {
            Log.e(logTag, "getTetheredIfaces invalid return type: $e")
            false
        } catch (e: Exception) {
            Log.e(logTag, "Error in getTetheredIfaces: $e")
            false
        }
    }

    object TetherMode {
        const val TETHERING_WIFI = 0
        const val TETHERING_USB = 1
        const val TETHERING_BLUETOOTH = 2
        const val TETHERING_NCM = 4
        const val TETHERING_ETHERNET = 5
        const val TETHERING_WIGIG = 6
        // VPNHotspot
        const val TETHERING_VPN = 7
    }

    /**
     * This enables tethering using the ssid/password defined in Settings App>Hotspot & tethering
     * Does not require app to have system/privileged access
     * Credit: Vishal Sharma - https://stackoverflow.com/a/52219887
     */
    @Suppress("unused")
    fun startTethering(mode: Int, callback: OnStartTetheringCallback?): Boolean {
        val outputDir: File = context.codeCacheDir
        val proxy: Any
        try {
            val callbackClass = onStartTetheringCallbackClass()
            if (callbackClass == null) {
                Log.e(logTag, "onStartTetheringCallbackClass returned null")
                return false
            }

            proxy = ProxyBuilder.forClass(callbackClass)
                .dexCache(outputDir).handler { proxy1, method, args ->
                    when (method.name) {
                        "onTetheringStarted" -> callback?.onTetheringStarted()
                        "onTetheringFailed" -> callback?.onTetheringFailed()
                        else -> ProxyBuilder.callSuper(proxy1, method, args)
                    }
                    //noinspection SuspiciousInvocationHandlerImplementation
                    null
                }.build()
        } catch (e: java.io.IOException) {
            Log.e(logTag, "ProxyBuilder IO error: $e")
            return false
        } catch (e: java.lang.reflect.InvocationTargetException) {
            Log.e(logTag, "ProxyBuilder invocation failed: $e")
            return false
        } catch (e: IllegalAccessException) {
            Log.e(logTag, "ProxyBuilder access denied: $e")
            return false
        } catch (e: InstantiationException) {
            Log.e(logTag, "ProxyBuilder instantiation failed: $e")
            return false
        } catch (e: Exception) {
            Log.e(logTag, "Error in enableTethering ProxyBuilder: $e")
            return false
        }

        return try {
            val callbackClass = onStartTetheringCallbackClass()
            if (callbackClass == null) {
                Log.e(logTag, "onStartTetheringCallbackClass returned null")
                return false
            }

            val method = connectivityManager.javaClass.getDeclaredMethod(
                "startTethering",
                Int::class.java,
                Boolean::class.java,
                callbackClass,
                Handler::class.java
            )
            method.invoke(connectivityManager, mode, false, proxy, null as Handler?)
            Log.d(logTag, "startTethering invoked")
            true
        } catch (e: NoSuchMethodException) {
            Log.e(logTag, "startTethering method not found: $e")
            false
        } catch (e: IllegalAccessException) {
            Log.e(logTag, "startTethering access denied: $e")
            false
        } catch (e: java.lang.reflect.InvocationTargetException) {
            Log.e(logTag, "startTethering invocation failed: $e")
            false
        } catch (e: Exception) {
            Log.e(logTag, "Error in enableTethering: $e")
            false
        }
    }

    fun stopTethering(mode: Int) {
        try {
            val method =
                connectivityManager.javaClass.getDeclaredMethod("stopTethering", Int::class.java)
            method.invoke(connectivityManager, mode)
            Log.d(logTag, "stopTethering invoked")
        } catch (e: Exception) {
            Log.e(logTag, "stopTethering error: $e")
        }
    }

    @SuppressLint("PrivateApi")
    private fun onStartTetheringCallbackClass(): Class<*>? {
        return try {
            Class.forName("android.net.ConnectivityManager\$OnStartTetheringCallback")
        } catch (e: ClassNotFoundException) {
            Log.e(logTag, "OnStartTetheringCallbackClass not found: $e")
            null
        } catch (e: Exception) {
            Log.e(logTag, "OnStartTetheringCallbackClass error: $e")
            null
        }
    }
}
